File: src/main/clojure/org/curious/pounce/body.clj
(comment
  Copyright 2010 Jarl Haggerty

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0
  
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.)

(ns org.curious.pounce.body
  "Defines the body structure and methods for simulating forces on it and calculating collisions."
  (:import java.awt.Color)
  (:require [org.curious.pounce.shape :as shape]
            [org.curious.pounce.math.core :as math]
            [org.curious.pounce.math.matrix :as matrix]))

(defn body= [this that] (and (= (:transformation this) (:transformation that))
                             (= (:shapes this) (:shapes that))
                             (math/eps= (:moment-of-inertia this) (:moment-of-inertia that))
                             (= (:linear-momentum this) (:linear-momentum that))
                             (= (:linear-velocity this)  (:linear-velocity that))
                             (math/eps= (:angular-momentum this)  (:angular-momentum that))
                             (math/eps= (:angular-velocity this)  (:angular-velocity that))
                             (math/eps= (:mass this)  (:mass that))
                             (= (:center-of-mass this)  (:center-of-mass that))
                             (= (:kinematic this)  (:kinematic that))))

(defn create [raw-trans & raw-shapes]
  "Creates a body from the specified transform and shapes."
  (let [[trans shapes] (if (instance? org.curious.pounce.math.matrix.Transformation raw-trans)
                         [raw-trans raw-shapes]
                         [matrix/identity-transformation (cons raw-trans raw-shapes)])
        mass (reduce + (map :mass shapes))
        center-of-mass (if (math/is-infinite mass)
                         (matrix/div (reduce matrix/add (map :center shapes)) (count shapes))
                         (matrix/div (reduce matrix/add (map #(matrix/mul (:center %) (:mass %)) shapes)) (count shapes)))]
    {:transformation trans
     :shapes shapes
     :moment-of-inertia (if (math/is-infinite mass)
			  math/positive-infinity
			  (reduce +
				  (map #(+ (:moment-of-inertia %)
					   (* (:mass %) (matrix/length-squared (matrix/sub (:center %) center-of-mass))))
				       shapes)))
     :linear-momentum (matrix/column 0 0)
     :linear-velocity (matrix/column 0 0)
     :angular-momentum 0
     :angular-velocity 0
     :mass mass
     :center-of-mass center-of-mass
     :kinematic (math/is-infinite mass)}))

(defn update [this delta force torque velocity]
  (let [new-linear-momentum (matrix/add (matrix/mul velocity (:mass this)) (:linear-momentum this) (matrix/mul force delta))
        new-angular-momentum (+ (:angular-momentum this) (* torque delta))
        new-linear-velocity (if (:kinematic this)
                              (:linear-velocity this)
                              (matrix/div new-linear-momentum (:mass this)))
        new-angular-velocity (if (:kinematic this)
                              (:angular-velocity this)
                              (/ new-angular-momentum (:moment-of-inertia this)))
        rotation-matrix (matrix/rotation-matrix (* delta new-angular-velocity))
        center (matrix/transform (:center-of-mass this) (:transformation this))]
    (assoc this
      :transformation (matrix/transformation (matrix/add (matrix/mul rotation-matrix
                                                                     (matrix/sub (-> this :transformation :translation)
                                                                                 center))
							 (matrix/mul new-linear-velocity delta)
                                                         center)
                                             (matrix/mul (-> this :transformation :rotation) rotation-matrix))
      :linear-velocity new-linear-velocity
      :angular-velocity new-angular-velocity
      :linear-momentum new-linear-momentum
      :angular-momentum new-angular-momentum)))

(defn render [this graphics] (doseq [shape (:shapes this)]
			       (shape/render (shape/transform shape (:transformation this)) graphics)))

File: src/main/java/org/curious/pounce/Body.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package org.curious.pounce;

import java.util.List;
import org.curious.pounce.math.Matrix;
import org.curious.pounce.math.Transformation;
import org.curious.pounce.math.Vector;

/**
 *
 * @author Jarl
 */
public class Body{
    public final Transformation transformation;
    public final List<Shape> shapes;
    public final double mass;
    public final double momentOfInertia;
    public final Vector center;
    public final Vector centerOfMass;
    public final Vector linearMomentum;
    public final double angularMomentum;

    public Body(Transformation transformation, List<Shape> shapes){
        this(transformation, shapes, new Vector(0, 0), 0);
    }

    public Body(Transformation transformation, List<Shape> shapes, Vector linearMomentum, double angularMomentum){
        this.transformation = transformation;
        this.shapes = shapes;

        double mass = 0;
        Vector.fill(Vector.reg0, 0, 0);

        for(Shape shape : shapes){
            mass += shape.mass;
            Vector.fill(Vector.reg1, shape.center);
            Vector.scale(Vector.reg1, shape.mass, Vector.reg1);
            Vector.add(Vector.reg0, Vector.reg1, Vector.reg0);
        }
        this.mass = mass;

        center = new Vector(0, 0);
        for(Shape shape : shapes){
            Vector.add(center, shape.center, center);
        }
        Vector.scale(center, 1/shapes.size(), center);

        if(mass == Double.POSITIVE_INFINITY){
            centerOfMass = center;
        }else{
            centerOfMass = new Vector(0, 0);
            for(Shape shape : shapes){
                Vector.fill(Vector.reg0, shape.center);
                Vector.scale(Vector.reg0, shape.mass, Vector.reg0);
                Vector.add(centerOfMass, Vector.reg0, centerOfMass);
            }
            Vector.scale(centerOfMass, 1/shapes.size(), centerOfMass);
        }

        double momentOfInertia = 0;
        for(Shape shape : shapes){
            Vector.sub(shape.center, centerOfMass, Vector.reg0);
            momentOfInertia += shape.momentOfIntertia + shape.mass*Vector.lengthSquared(Vector.reg0);
        }
        this.momentOfInertia = momentOfInertia;

        this.linearMomentum = linearMomentum;
        this.angularMomentum = angularMomentum;
    }

    public Body(Body body, double delta, Vector force, double torque){
        shapes = body.shapes;
        mass = body.mass;
        center = body.center;
        centerOfMass = body.centerOfMass;
        momentOfInertia = body.momentOfInertia;
        linearMomentum = new Vector(force);
        Vector.scale(linearMomentum, delta, linearMomentum);
        Vector.add(linearMomentum, body.linearMomentum, linearMomentum);
        angularMomentum = torque*delta + body.angularMomentum;

        Matrix rotation = Matrix.rotationMatrix(angularMomentum*delta/momentOfInertia);
        Matrix.mul(body.transformation.rotation, rotation, rotation);

        Vector translation = new Vector(linearMomentum);
        Vector.scale(translation, delta/mass, translation);
        Vector.add(translation, body.transformation.translation, translation);

        Vector.sub(body.transformation.translation, body.centerOfMass, Vector.reg0);
        Matrix.mul(body.transformation.rotation, Vector.reg0, Vector.reg1);
        Vector.sub(Vector.reg1, Vector.reg0, Vector.reg0);
        Vector.add(translation, Vector.reg0, translation);

        transformation = new Transformation(rotation, translation);
    }
}
